OrbitCaptureClient/CaptureEventProcessorTest.cpp

```diff
@@ -0,0 +1,614 @@
+using orbit_grpc_protos::ThreadStateSlice;
+using orbit_grpc_protos::TracepointEvent;
+using orbit_grpc_protos::TracepointInfo;
+
+using ::testing::_;
+using ::testing::DoAll;
+using ::testing::SaveArg;
+
+namespace {
+
+class MockCaptureListener : public CaptureListener {
+ public:
+  MOCK_METHOD(
+      void, OnCaptureStarted,
+      (ProcessData&& /*process*/,
+       (absl::flat_hash_map<uint64_t, orbit_client_protos::FunctionInfo>)/*selected_functions*/,
+       TracepointInfoSet /*selected_tracepoints*/,
+       UserDefinedCaptureData /*user_defined_capture_data*/),
+      (override));
+  MOCK_METHOD(void, OnCaptureComplete, (), (override));
+  MOCK_METHOD(void, OnCaptureCancelled, (), (override));
+  MOCK_METHOD(void, OnCaptureFailed, (ErrorMessage), (override));
+  MOCK_METHOD(void, OnTimer, (const TimerInfo&), (override));
+  MOCK_METHOD(void, OnKeyAndString, (uint64_t /*key*/, std::string), (override));
+  MOCK_METHOD(void, OnUniqueCallStack, (CallStack), (override));
+  MOCK_METHOD(void, OnCallstackEvent, (CallstackEvent), (override));
+  MOCK_METHOD(void, OnThreadName, (int32_t /*thread_id*/, std::string /*thread_name*/), (override));
+  MOCK_METHOD(void, OnThreadStateSlice, (ThreadStateSliceInfo), (override));
+  MOCK_METHOD(void, OnAddressInfo, (LinuxAddressInfo), (override));
+  MOCK_METHOD(void, OnUniqueTracepointInfo, (uint64_t /*key*/, TracepointInfo /*tracepoint_info*/),
+              (override));
+  MOCK_METHOD(void, OnTracepointEvent, (TracepointEventInfo), (override));
+};
+
+TEST(CaptureEventProcessor, CanHandleSchedulingSlices) {
+  MockCaptureListener listener;
+  CaptureEventProcessor event_processor(&listener);
+
+  CaptureEvent event;
+  SchedulingSlice* scheduling_slice = event.mutable_scheduling_slice();
+  scheduling_slice->set_core(2);
+  scheduling_slice->set_pid(42);
+  scheduling_slice->set_tid(24);
+  scheduling_slice->set_in_timestamp_ns(3);
+  scheduling_slice->set_out_timestamp_ns(100);
+
+  TimerInfo actual_timer;
+  EXPECT_CALL(listener, OnTimer).Times(1).WillOnce(SaveArg<0>(&actual_timer));
+
+  event_processor.ProcessEvent(event);
+
+  EXPECT_EQ(actual_timer.start(), scheduling_slice->in_timestamp_ns());
+  EXPECT_EQ(actual_timer.end(), scheduling_slice->out_timestamp_ns());
+  EXPECT_EQ(actual_timer.process_id(), scheduling_slice->pid());
+  EXPECT_EQ(actual_timer.thread_id(), scheduling_slice->tid());
+  EXPECT_EQ(actual_timer.processor(), scheduling_slice->core());
+  EXPECT_EQ(actual_timer.type(), TimerInfo::kCoreActivity);
+}
+
+TEST(CaptureEventProcessor, CanHandleOneCallstackSample) {
+  MockCaptureListener listener;
+  CaptureEventProcessor event_processor(&listener);
+
+  CaptureEvent event;
+  CallstackSample* callstack_sample = event.mutable_callstack_sample();
+  callstack_sample->set_pid(1);
+  callstack_sample->set_tid(3);
+  Callstack* callstack = callstack_sample->mutable_callstack();
+  callstack->add_pcs(14);
+  callstack->add_pcs(15);
+  callstack_sample->set_timestamp_ns(100);
```

Comment:
Have you considered putting that into a free static function since this function could be re-used in the next test?

Reply:
Thanks, I created an init function.